File: app/javascript/channels/admin_program_channel.ts
import { Program } from "../lib/types/types";
import consumer from "./consumer";

export interface AdminProgramChannel {
  setDebounceInterval: (interval: number) => void;
  setMode: (mode: "Anarchy" | "Democracy") => void;
  setMaxInputMode: (mode: number) => void;
  setVoteInterval: (interval: number) => void;
  setVoteThreshold: (threshold: number) => void;
  setCanVote: (value: boolean) => void;
  setConfetti: (value: boolean) => void;
  pause: () => void;
  resume: () => void;
  clear: () => void;
  reset: () => void;
};

export const AdminProgramChannel = (program: Program, setProgram: (program: Program) => void): AdminProgramChannel => {
  const updateProgram = (newProgram) => {
    program = newProgram;
    setProgram(program);
  };

  const sub = consumer.subscriptions.create(
    { channel: "AdminProgramChannel", id: program.id },
    {
      connected() {
        console.debug(`Connected to Admin Program Channel(${program.id})`);
      },

      disconnected() {
        console.debug("disconnected");
        setProgram({} as Program);
      },

      received({ action, data }) {
        switch (action) {
          case "tick":
            console.debug("tick");
            updateProgram({ ...program, ...data });
            break;

          case "output":
            updateProgram({ ...program, ...{ output: data } });
            break;

          default:
            updateProgram({ ...program, ...data });
            break;
        }
      },
    }
  );
  return {
    setDebounceInterval: (interval: number) => sub.perform("set_debounce_interval", { data: interval }),
    setMode: (mode: "Anarchy" | "Democracy") => sub.perform("set_mode", { data: mode }),
    setMaxInputMode: (mode: number) => sub.perform("set_max_input_mode", { data: mode }),
    setVoteInterval: (interval: number) => sub.perform("set_vote_interval", { data: interval }),
    setVoteThreshold: (threshold: number) => sub.perform("set_vote_threshold", { data: threshold }),
    setCanVote: (value) => sub.perform("set_can_vote", { data: value }),
    setConfetti: (value) => sub.perform("set_confetti", { data: value }),
    pause: () => sub.perform("pause"),
    resume: () => sub.perform("resume"),
    clear: () => sub.perform("clear"),
    reset: () => {
      sub.perform("clear");
    },
  };
};

File: app/channels/admin_program_channel.rb
class AdminProgramChannel < ApplicationCable::Channel
  periodically :tick, every: 5.seconds

  def subscribed
    stream_for admin_room
  end

  def unsubscribed
  end

  def tick
    AdminProgramChannel.broadcast_to(admin_room, {
      action: :tick,
      data: current_program.tick_view
    })
  end

  def set_debounce_interval(message)
    current_program.update(settings: current_program.settings.merge({
      debounce_interval: message["data"]
    }))
    ProgramChannel.broadcast_to(room, {
      action: :tick,
      data: current_program.tick_view
    })
  end

  def set_mode(message)
    current_program.update(
      mode: message["data"],
      settings: current_program.settings.merge({
        vote_interval: Program::VOTE_THRESHOLD[message["data"]],
        vote_threshold: Program::VOTE_THRESHOLD[message["data"]],
      })
    )
    ProgramChannel.broadcast_to(room, {
      action: :tick,
      data: current_program.tick_view
    })
  end

  def set_max_input_mode(message)
    current_program.update(settings: current_program.settings.merge({
      max_input_mode: message["data"]
    }))
    ProgramChannel.broadcast_to(room, {
      action: :tick,
      data: current_program.tick_view
    })
  end

  def set_can_vote(message)
    current_program.update(settings: current_program.settings.merge({
      can_vote: message["data"]
    }))
    ProgramChannel.broadcast_to(room, {
      action: :tick,
      data: current_program.tick_view
    })
  end

  def set_vote_interval(message)
    current_program.update(settings: current_program.settings.merge({
      vote_interval: message["data"]
    }))
    ProgramChannel.broadcast_to(room, {
      action: :tick,
      data: current_program.tick_view
    })
  end

  def set_vote_threshold(message)
    current_program.update(settings: current_program.settings.merge({
      vote_threshold: message["data"]
    }))
    ProgramChannel.broadcast_to(room, {
      action: :tick,
      data: current_program.tick_view
    })
  end

  def set_confetti(message)
    current_program.update(settings: current_program.settings.merge({
      confetti: message["data"]
    }))
    ProgramChannel.broadcast_to(room, {
      action: :tick,
      data: current_program.tick_view
    })
  end

  def pause
    current_program.update(settings: current_program.settings.merge({
      play_state: "paused" }))
    ProgramChannel.broadcast_to(room, {
      action: :tick,
      data: current_program.tick_view
    })
  end

  def resume
    current_program.update(settings: current_program.settings.merge({
      play_state: "playing" }))
    ProgramChannel.broadcast_to(room, {
      action: :tick,
      data: current_program.tick_view
    })
  end

  def receive(data)
    puts "receive:#{data}"
  end

  def clear
    program = current_program
    program.entries.destroy_all
    program.chars.destroy_all

    ProgramChannel.broadcast_to(room, {
      action: :clear,
      data: program.view
    })
  end

  private
  def current_program
    Program.includes(:entries, :messages).find(params.fetch(:id))
  end

  def room
    "program_#{current_program.id}"
  end

  def admin_room
    "admin_program_#{current_program.id}"
  end
end
